lyse/__init__.py

```diff
@@ -296,6 +299,102 @@ def h5_path(self):
     def no_write(self):
         """bool: The value provided for `no_write` during instantiation."""
         return self.__no_write
+    
+    @property
+    def h5_file(self):
+        """h5py.File: opened h5py file handle for the shot"""
+        return self.__h5_file
+    
+    @contextlib.contextmanager
+    def open(self, mode):
+        """Opens the Run's h5 file for successive reads/writes.
+
+        Intended to be used as a context manager.
+        Supports all h5 modes, though only `'r'` and `'r+'`/`'a'` are typically
+        used within lyse itself.
+
+        Args:
+            mode (str): which mode to open the h5 file with.
+                Must be 'r', 'a', 'r+', 'w', 'w-', or 'x'.
+                Lyse typically only uses 'r' and 'a'.
+
+        Examples:
+            >>> from lyse import *
+            >>> shot = Run(path)
+            >>> with shot.open('a'):
+            >>>     # shot processing that requires reads/writes
+            >>>     t, vals = shot.get_trace('my_trace')
+            >>>     results = vals**2
+            >>>     shot.save_result_array('my_result', results)
+            >>> # Shot processing that doesn't require h5 reads/writes
+            >>> print(f'Max of results is {results.max():.3f}')
+            Max of results is 5.003
+
+        Raises:
+            PermissionError: If the Run is set as read-only but a write mode is requested
+        """
+
+        # ensure no_write is respected
+        if self.no_write:
+            if mode == 'r':
+                pass
+            else:
+                msg = f'Cannot open file with a write mode; this run is read-only'
+                raise PermissionError(msg)
+            
+        # ensure file is not already opened
+        if self.__h5_file is not None:
+            msg = 'h5_file already opened'
+            raise PermissionError(msg)
```

Comment:
If we new the mode wasn’t changing, there is nothing stopping us yielding `self.__h5_file` here (and not closing the file afterwards). Does that go some way to helping with resuable code snippets that might not know if they’re being called inside the context manager already?

Reply:
Fair enough. Just need to make sure the modes are compatible, then return.

Reply:
Actually, I don't think we have to yield the file handle (normal operation doesn't). If fact, I think it has to `return None` here to match signatures and prevent executing the file open, right?

Reply:
You need to prevent the later open and yield occurring. I think you still want to use yield too though. Maybe:

```
    yield self.__h5_file
    return
```

Otherwise you don't get the file handle out the second time, so you get incompatible behaviour of you reuse some code inside an already open context manager. And I don't know how the `contextmanager` decorator handles no yield.

Reply:
Well I can report that contextManager doesn't like it if there isn't a yield since it is a generator. TIL.

I'm still confused why it should yield the file handle instead of `None`. The file handle isn't returned by this context normally; it just saves it as a Run attribute.

I suppose that means we can't do `with run.open('r') as f`, but my working assumption is that we didn't want people to normally access the file directly. If they really need the handle, they can get it from the `run.h5_file` property. I think I slightly prefer that (as it discourages casual direct interaction with the h5 file since you need to separately get the handle) but I don't have strong feelings about it. Either is fine.

Reply:
The reason to yield the file handle is so it mimics `h5py.File(..)`. Probably most people won't use it. And yep, you could access it via the attribute on the `Run` instance. But given we have to yield, we may as well yield something useful! You don't have to even assign it (`with run.open(mode)` should work as well as `with run.open(mode) as f`)

Reply:
Think I must have not read the whole comment before I replied. Feel like I've just repeated what you said. Oops. I don't see any harm in given people the file handle though. Sometimes you might want direct access to the file. Generally you expect context managers that open a file to yield the file handle as well. I think my feelings are mainly driven by that - trying to be consistent with how you'd expect it to behave.